contracts/TreasuryInsurance.sol

```diff
@@ -0,0 +1,100 @@
+import '@openzeppelin/contracts/utils/Context.sol';
+import '@openzeppelin/contracts/token/ERC20/IERC20.sol';
+import '@openzeppelin/contracts/token/ERC721/ERC721.sol';
+import './interfaces/IERC20Allowance.sol';
+import './Treasury.sol';
+
+// Treasury contract with insurance.
+contract TreasuryInsurance is Treasury, ERC721 {
+    mapping(uint256 => uint256) public compensationLimits;
+    mapping(uint256 => string) public insuranceConditions;
+    mapping(uint256 => uint256) public insuranceCosts;
+    mapping(uint256 => address) public insuranceAssets;
+    mapping(uint256 => uint256) public lastPayTime;
+    mapping(uint256 => uint256) public requestedCompensations;
+    mapping(uint256 => string) public requestedCases;
+
+    uint256 public insuranceCounter;
+
+    event Insured(
+        uint256 id,
+        address insured,
+        address asset,
+        uint256 cost,
+        uint256 compensationLimit
+    );
+    event Requested(uint256 id, uint256 compensation);
+    event PaidInsurance(uint256 id, uint256 payment);
+    event Compensated(uint256 id, address insured, uint256 compensationLimit);
+
+    constructor(address governance_, address executor_)
+        Treasury(governance_, executor_)
+        ERC721('BITDAO_TREASURY_INSURANCE', 'BITTI')
+    {}
+
+    function insure(
+        address to,
+        address asset,
+        uint256 cost,
+        uint256 compensationLimit,
+        string memory condition
+    ) external virtual onlyExecutor returns (uint256) {
+        compensationLimits[insuranceCounter] = compensationLimit;
+        insuranceConditions[insuranceCounter] = condition;
+        insuranceCosts[insuranceCounter] = cost;
+        insuranceAssets[insuranceCounter] = asset;
+        lastPayTime[insuranceCounter] = block.timestamp;
+        _safeMint(to, insuranceCounter);
+        insuranceCounter += 1;
+        emit Insured(insuranceCounter, to, asset, cost, compensationLimit);
+        return insuranceCounter;
+    }
+
+    function payInsurance(uint256 id, uint256 period) public virtual {
+        lastPayTime[id] = block.timestamp + period;
+        IERC20(insuranceAssets[id]).transferFrom(
+            _msgSender(),
+            address(this),
+            period * insuranceCosts[id]
+        );
+        emit PaidInsurance(id, period * insuranceCosts[id]);
+    }
+
+    function requestInsurance(
+        uint256 id,
+        uint256 compensation,
+        string memory reason
+    ) external virtual {
+        require(
+            ownerOf(id) == _msgSender(),
+            'Insurance::request: requester must be owner'
+        );
+        require(
+            compensation <= compensationLimits[id],
+            'Insurance::request: requested compensation is more than the limit'
+        );
+        requestedCompensations[id] = compensation;
+        requestedCases[id] = reason;
+        if (block.timestamp > lastPayTime[id]) {
+            payInsurance(id, block.timestamp - lastPayTime[id]);
```

Comment:
The invocation of "payInsurance()" looks weird to me.

insure() -> admin issue an insurance to the beneficiary
payInsurance() -> beneficiary pays a fee for the period he select
requestInsurance() -> this is confusing, it can be understood as beneficiary request for pay for an insurance, can be also understood as beneficiary request to claim an insurance

That's why I recommend to add documentation everywhere to convey your intention.

To me the code looks more like a beneficiary want to claim an insurance with a reason, so we update his requestedCompensation map. If so, there is no point for the beneficiary to invoke the "payInsurance()" inside this function, because the payInsurance is one directional payment flow from user to this contract. The contract should pay out to him actually.

Reply:
For insurance request, Governance needs to process this and make compensation separately, users though should pay insurance up to this point of request and so calling payInsurance insures he did that and easier and cheaper to use for users, you just pay for everything before the request.